pkg/aggregateeventrecorder/aggregateeventrecorder.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/ipfs/go-log/v2"
+)
+
+var logging = log.Logger("aggregateeventrecorder")
+
+const httpTimeout = 5 * time.Second
+const parallelPosters = 5
+
+type tempData struct {
+	startTime time.Time
+	ttfb      float64
+}
+
+type AggregatedEvent struct {
+	InstanceID        string    `json:"instanceId"`        // The ID of the Lassie instance generating the event
+	RetrievalID       string    `json:"retrievalId"`       // The unique ID of the retrieval
+	StorageProviderID string    `json:"storageProviderId"` // The ID of the storage provider that served the retrieval content
+	TimeToFirstByte   float64   `json:"timeToFirstByte"`   // The time it took to receive the first byte in seconds
+	Bandwidth         float64   `json:"bandwidth"`         // The bandwidth of the retrieval in bytes per second
+	Success           bool      `json:"success"`           // Wether or not the retreival ended with a success event
+	StartTime         time.Time `json:"startTime"`         // The time the retrieval started
+	EndTime           time.Time `json:"endTime"`           // The time the retrieval ended
+}
+
+type batchedEvents struct {
+	Events []AggregatedEvent `json:"events"`
+}
+
+// NewAggregateEventRecorderSubsciber returns a RetrievalEventSubscriber that POSTs
+// a batch of aggregated retrieval events to an event recorder API endpoint
+func NewAggregateEventRecorderSubscriber(
+	ctx context.Context,
+	instanceID string,
+	endpointURL string,
+	endpointAuthorization string,
+	disableIndexerEvents bool,
+) types.RetrievalEventSubscriber {
+
+	incomingChan := make(chan types.RetrievalEvent)
+	postChan := make(chan []AggregatedEvent)
+	eventTempMap := make(map[string]tempData)
+
+	// Ingest data
+	go func() {
+		var batchedData []AggregatedEvent
+		var emptyGaurdChan chan []AggregatedEvent = nil
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			// read incoming data
+			case event := <-incomingChan:
+				id := event.RetrievalId().String()
+
+				switch event.Code() {
+				case types.StartedCode:
+					// Save the start time so we can calculate time to first byte
+					eventTempMap[id] = tempData{event.Time(), 0}
+
+				case types.FirstByteCode:
+					// Calculate time to first byte
+					tempData := eventTempMap[id]
+					tempData.ttfb = event.Time().Sub(tempData.startTime).Seconds()
+					eventTempMap[id] = tempData
+
+				case types.SuccessCode, types.FailedCode:
+					tempData := eventTempMap[id]
+
+					success := event.Code() == types.SuccessCode
+					spId := event.StorageProviderId().String()
+					endTime := event.Time()
+
+					var bytesPerSec float64
+					if success {
+						byts := event.(events.RetrievalEventSuccess).ReceivedSize()
+						duration := event.(events.RetrievalEventSuccess).Duration().Seconds()
+						bytesPerSec = float64(byts) / duration
+					}
+
+					aggregatedEvent := AggregatedEvent{
+						InstanceID:        instanceID,
+						RetrievalID:       id,
+						StorageProviderID: spId,
```

Comment:
got a tricky decision to make on what to do with this on the failure case - most retrievals are going to have multiple failures with different SPs but one success with a single SP. The docs for this struct say it's the SP ID for the successful retrieval, so do we just ignore all the ones that failed? in that case, `spId` needs to be empty if we never get a success

Reply:
yea so, this needs some discussion. I am thinking about whether we need metrics on failures for SPs -- think if we had to serve the golden retriever query.

also, we need some information on the farthest point we reached as well.

Reply:
I've made SP ID an empty string if we get a Finish event and no SP ID has been saved from a Success event.